### ISSUE TITLE
Parallel use of prefetch of SRA files from AWS S3 leads to NCBI abuse message & blocking?

### ISSUE THREAD

**Issue by taltman** [ncbi/ncbi-vdb#28]
Hi all,

My apologies if this is not the right place to file this issue.

I'm downloading SRA files using prefetch on an AWS EC2 instance in the
us-east-1 region. I've used vdb-config -i to make sure that I'm using
AWS S3 to fetch the SRA files. Here's the output of srapath to show that I'm fetching from S3:

```
> srapath ERR169824
https://sra-pub-run-5.s3.amazonaws.com/ERR169824/ERR169824.1
```

I'm on a beefy EC2 instance with 96 cores, with each core running prefetch and then processing the SRA file with blast_vdb. Sometimes the two processes complete in a few seconds, sometimes it takes them up to 4 minutes. After my job was running for a while, I got this message:

```
Critical: (307.10) [NCBI-MESSAGE]  [DISPD#ID2@sutils201] Auto-blocked
for abuse, please contact NCBI help desk <info@ncbi.nlm.nih.gov>
```

Are there resource limitations when downloading on S3, as there are with
using EUtils? Time-of-day, time-of-week, and/or rate limits? How to avoid this issue?

Thank you!

**Comment by kwrodarmer** [comment]
Please use the latest version of the toolkit. It appears you are using version 2.10.0. The latest is 2.10.7.

**Comment by taltman** [comment]
I can confirm that I am using version 2.10.7:
```
/sratoolkit.2.10.7-ubuntu64/bin/prefetch -V

"./sratoolkit.2.10.7-ubuntu64/bin/prefetch" version 2.10.7
```

**Comment by kwrodarmer** [comment]
Okay. Your original report has
```
> srapath ERR169824
```

I can't tell what the version of `srapath` is there. If you're certain it's 2.10.7, then we'll start investigating what could cause the behavior you're seeing. Are you certain that you do not have 2.10.0 on your system and in your $PATH?

**Comment by taltman** [comment]
iftop is showing me the following when I download:

```
ip-172-31-72-183.ec2.internal                                                                         => s3-1-w.amazonaws.com                                                                                      0b    227b    101b
                                                                                                      <=                                                                                                           0b      0b      0b
ip-172-31-72-183.ec2.internal                                                                         => srafiles11-1.be-md.ncbi.nlm.nih.gov                                                                       0b      0b   81.1Kb
                                                                                                      <=                                                                                                           0b      0b   20.8Mb 
ip-172-31-72-183.ec2.internal                                                                         => srafiles33-1.st-va.ncbi.nlm.nih.gov                                                                       0b      0b   80.1Kb
                                                                                                      <=                                                                                                           0b      0b   17.9Mb
ip-172-31-72-183.ec2.internal                                                                         => srafiles12-1.be-md.ncbi.nlm.nih.gov                                                                       0b      0b   56.6Kb
                                                                                                      <=                                                                                                           0b      0b   17.1Mb
ip-172-31-72-183.ec2.internal                                                                         => srafiles21-1.be-md.ncbi.nlm.nih.gov                                                                       0b      0b   69.2Kb
                                                                                                      <=                                                                                                           0b      0b   15.8Mb
ip-172-31-72-183.ec2.internal                                                                         => srafiles13-1.be-md.ncbi.nlm.nih.gov                                                                       0b      0b   27.4Kb
```

So in addition to hitting S3 servers, there are some connections being made to NCBI.

Here's what prefetch is emitting to the terminal:

```
2020-06-17T20:07:23 prefetch.2.10.7: 129) 'ncbi-acc:NC_000020.11?vdb-ctx=refseq' was downloaded successfully
2020-06-17T20:07:23 prefetch.2.10.7: 130) Downloading 'ncbi-acc:NC_000021.9?vdb-ctx=refseq'...
2020-06-17T20:07:23 prefetch.2.10.7:  Downloading via HTTPS...
2020-06-17T20:07:24 prefetch.2.10.7:  HTTPS download succeed
2020-06-17T20:07:24 prefetch.2.10.7: 130) 'ncbi-acc:NC_000021.9?vdb-ctx=refseq' was downloaded successfully
2020-06-17T20:07:24 prefetch.2.10.7: 131) Downloading 'ncbi-acc:NC_000022.11?vdb-ctx=refseq'...
2020-06-17T20:07:24 prefetch.2.10.7:  Downloading via HTTPS...
2020-06-17T20:07:25 prefetch.2.10.7:  HTTPS download succeed
2020-06-17T20:07:25 prefetch.2.10.7: 131) 'ncbi-acc:NC_000022.11?vdb-ctx=refseq' was downloaded successfully
2020-06-17T20:07:25 prefetch.2.10.7: 132) Downloading 'ncbi-acc:NC_000023.11?vdb-ctx=refseq'...
2020-06-17T20:07:25 prefetch.2.10.7:  Downloading via HTTPS...
2020-06-17T20:07:27 prefetch.2.10.7:  HTTPS download succeed
2020-06-17T20:07:27 prefetch.2.10.7: 132) 'ncbi-acc:NC_000023.11?vdb-ctx=refseq' was downloaded successfully
2020-06-17T20:07:27 prefetch.2.10.7: 133) Downloading 'ncbi-acc:NC_000024.10?vdb-ctx=refseq'...
2020-06-17T20:07:27 prefetch.2.10.7:  Downloading via HTTPS...
2020-06-17T20:07:28 prefetch.2.10.7:  HTTPS download succeed
2020-06-17T20:07:28 prefetch.2.10.7: 133) 'ncbi-acc:NC_000024.10?vdb-ctx=refseq' was downloaded successfully
2020-06-17T20:07:28 prefetch.2.10.7: 134) Downloading 'ncbi-acc:NC_012920.1?vdb-ctx=refseq'...
2020-06-17T20:07:28 prefetch.2.10.7:  Downloading via HTTPS...
2020-06-17T20:07:28 prefetch.2.10.7:  HTTPS download succeed
2020-06-17T20:07:28 prefetch.2.10.7: 134) 'ncbi-acc:NC_012920.1?vdb-ctx=refseq' was downloaded successfully
```

So I guess the question is why is prefetch still querying NCBI, even when I have things set up to download from S3? Do I have something misconfigured?

**Comment by kwrodarmer** [comment]
Can you run
```
test-sra ERR169824 > test-sra-out.txt
```

and send the output file to sra-tools@ncbi.nlm.nih.gov, please?

**Comment by taltman** [comment]
Done! Thanks!

**Comment by kwrodarmer** [comment]
Once you send the output, I will ask for it to be forwarded for examination.

Can you tell how old your configuration is? This is a loaded question, because our configuration is tiered. We look in `/etc/ncbi` first, then along side the binary there is some default configuration, then under $HOME. If there is anything in that chain that could be pointing you to an old server, i.e. a holdover from older installations, that could potentially cause some issues.

I see you sent the file - I'll ask it to be forwarded ASAP. Thanks, and very sorry for the trouble.

**Comment by kwrodarmer** [comment]
Okay, thank you very much for sending the file. We have examined it and cannot find any issues in your configuration. We suspect there may be an issue in our server backend here. Engineers are following that hypothesis now.

We'll follow up as soon as we learn anything, and again apologies for the trouble.

**Comment by taltman** [comment]
My config info:
```
(base) ubuntu@ip-172-31-72-183:~/.ncbi$ ls -lh
total 4.0K
-rw------- 1 ubuntu ubuntu 701 Jun 12 18:44 user-settings.mkfg
(base) ubuntu@ip-172-31-72-183:/etc$ ls /etc/ncbi
ls: cannot access '/etc/ncbi': No such file or directory
```

**Comment by taltman** [comment]
> You might try configuring so that you do not accept charges, but accept delivery of compute environment token.

I'm not sure I understand the steps here. I know how to use `vdb-config -i` to toggle the "accept charges" option for AWS. But the environment token I don't know about. Could you please break it down into specific steps for me? Thanks in advance!

**Comment by kwrodarmer** [comment]
It's
```
[ ] accept charges for AWS
[x] report cloud instance identity
credentials
[choose] ____________________
[clear]
profile: _________________
```

**Comment by taltman** [comment]
What is the option to "accept delivery of compute environment token"? That's the part that is unclear to me.

**Comment by taltman** [comment]
Also, could you please briefly describe how this work-around works? I thought that selecting "accept charges for AWS" was a prerequisite for accessing SRA files from AWS S3?

**Comment by kwrodarmer** [comment]
I gave the exact wording above:

> report cloud instance identity

That's what you want to check, and uncheck the one marked

> accept charges for AWS

if you're working within us-east-1. That's because there won't be any egress charges from there, so as long as you agree to report your "cloud instance identity" which is a token from AWS that proves your location, then we don't need payment credentials for operation. I'm suggesting you try this to see if anything changes.

**Comment by kwrodarmer** [comment]
To be more clear, the general case for accessing data from AWS is that you need to supply payment credentials to AWS bucket servers in order to access data. You configure the toolkit to send those credentials on your behalf when you "accept charges" and provide a pointer to your credentials.

But as an optimization, when you can show you are in a compatible region of the cloud, we can give you access that does not require payment credentials because there will be no egress from the zone. When you "report cloud instance identity" it can tell us that you're in one of those special regions and we generate a different bucket access than otherwise; one that doesn't depend upon credentials.

**Comment by taltman** [comment]
Ok. I just tried that configuration, and I got the same error message as before. 

Here's a more minimal reproducible example, which doesn't involve multiple cores:

I remove the locally-downloaded SRA file, and I grab it again using prefetch:
```
(base) ubuntu@ip-172-31-72-183:/dev/shm$ rm -r SRR11647688
(base) ubuntu@ip-172-31-72-183:/dev/shm$ prefetch SRR11647688

2020-06-18T00:20:47 prefetch.2.10.7: 1) Downloading 'SRR11647688'...
2020-06-18T00:20:47 prefetch.2.10.7:  Downloading via HTTPS...
2020-06-18T00:21:45 prefetch.2.10.7:  HTTPS download succeed
2020-06-18T00:21:45 prefetch.2.10.7: 1) 'SRR11647688' was downloaded successfully
2020-06-18T00:21:45 prefetch.2.10.7: 'SRR11647688' has 0 unresolved dependencies
```

I validate the SRA file:
```
(base) ubuntu@ip-172-31-72-183:/dev/shm$ vdb-validate SRR11647688/SRR11647688.sra 
2020-06-18T00:22:56 vdb-validate.2.10.7 info: Database 'SRR11647688.sra' metadata: md5 ok
2020-06-18T00:22:56 vdb-validate.2.10.7 info: Table 'SEQUENCE' metadata: md5 ok
2020-06-18T00:22:56 vdb-validate.2.10.7 info: Column 'ALTREAD': checksums ok
2020-06-18T00:22:56 vdb-validate.2.10.7 info: Column 'QUALITY': checksums ok
2020-06-18T00:22:56 vdb-validate.2.10.7 info: Column 'READ': checksums ok
2020-06-18T00:23:00 vdb-validate.2.10.7 warn: type unrecognized while validating database - Database 'SRR11647688/SRR11647688.sra' has unrecognized type 'NCBI:SRA:GenericFastq:db'
2020-06-18T00:23:00 vdb-validate.2.10.7 info: Database 'SRR11647688.sra' is consistent
```

I try to run `blastn_vdb` on the SRA file directly, but it claims that there is something missing from the file
```
(base) ubuntu@ip-172-31-72-183:/dev/shm$ time blastn_vdb -db SRR11647688/SRR11647688.sra -query gene.fna > test.tsv
Error: Cannot get VDB column: /dev/shm/SRR11647688/SRR11647688.sra.SEQUENCE.ACCESSION
```

So I try running `blastn_vdb` again, using the SRA accession alone, and not the relative path to the *.sra file:
```(base) ubuntu@ip-172-31-72-183:/dev/shm$ time blastn_vdb -db SRR11647688 -query DEspR.fna > test.tsv
Critical: (307.10) [NCBI-MESSAGE]  [DISPD#ID2@sutils201] Auto-blocked for abuse, please contact NCBI help desk <info@ncbi.nlm.nih.gov>
Error: (301.5) [CONN_Open(ID2; https://www.ncbi.nlm.nih.gov/Service/dispd.cgi)]  Failed to open connection: Unknown
CReader: cannot open initial connection: NCBI C++ Exception:
    T0 "/panfs/pan1/sra-test/vdb3/git/blast_vdb/sratoolkit_2_9_0/c++/src/objtools/data_loaders/genbank/id2/reader_id2.cpp", line 257: Error: ncbi::objects::CId2Reader::x_ConnectAtSlot() - cannot open connection: ID2
```

This causes the error message.

So my questions are:

1. Why does `blastn_vdb` report something malformed about the SRA file when given the relative path?
2. If `prefetch` works fine, and the error is with `blastn_vdb`, then is something misconfigured with my cache, thus prompting `blastn_vdb` to try to download the SRA from NCBI?

**Comment by kwrodarmer** [comment]
We're looking into it. We closed 5 hours ago, and will continue tomorrow.

**Comment by taltman** [comment]
Great, thanks for all of your help!

In case it is relevant, I'm also seeing that `cache-mgr` is not working; more evidence that this bug has something to do with local caching?

```
(base) ubuntu@ip-172-31-72-183:/dev/shm$ cache-mgr -r
2020-06-18T06:43:57 cache-mgr.2.10.7 err: path not found while opening node within configuration module - KRepositoryRoot( for repo # 0 ) failed in get_cache_path_from_repo_mgr
```

**Comment by kwrodarmer** [comment]
Our `2.10` release had more changes under the hood than probably any other release we've ever had, and there are tools that are still catching up.

`cache-mgr` is a tool for managing dedicated caches, and our caching scheme changed in a way that I'll try to describe. The fact that there was an error reported means we need to update the tool to give a meaningful answer, but there's a reason for the error.

Before `2.10`, caching was on or off. When on, we cached in a user-designated area that mimicked a mini-repository. By default this was under the user's $HOME (where else could be default?), although users were encouraged to select an appropriate place. All caching happened there, so `prefetch` would download there and `fasterq-dump` would read from there, and all download-on-demand was stored there. While not being simple, it was simplistic.

Starting with `2.10`, we shifted our default behavior in large part because of parallel usage. Download-on-demand now creates temp files by default, and `prefetch` downloads to an "accession-directory". The temporary files are removed when the tool exits, and the accession-directory is not considered part of any cache that would be managed by `cache-mgr`.

It's still possible to set up an older-style user-cache area, and we honor them if they are found. If it exists, `cache-mgr` will find it and not report an error. The idea behind `cache-mgr` was to be able to see and potentially clean up files that were being hidden from view. The accession-directory is under user management, and temp files are automatically cleaned up.

So that's what's going on with cache and `cache-mgr`. We'll create a ticket to improve intelligibility of its output.

**Comment by kwrodarmer** [comment]
As we start off the day, I do want to make a comment about the error you reported initially; NCBI abuse messages and blocking.

The SRA Toolkit makes use of a name resolver service at NCBI called the "SDL" or "SRA Data Locator." Since `2.10` we are making more use of it than ever because we now have triple the number of possible data locations and don't want to place the logic for location in the tools.

The SDL, in turn, makes use of various databases within NCBI in order to identify and locate run data. It appears so far that the SDL is being blocked by other internal systems that were originally designed to support human-speed interaction. These systems generate errors under ancient assumptions that are reported to SDL which reports them to the SRA Toolkit which displays them on the screen.

There is nothing in the design of the SRA Toolkit or the SRA repository that should be applying such restrictions. So we are working on tracking this down and fixing it.

**Comment by kwrodarmer** [comment]
@taltman - can you try an experiment?

Please set this in your environment:
```
$NCBI_CONFIG__GENBANK__PREOPEN=false
```

and run `blastn_vdb` with that setting, and see if this addresses your rate-limit issue.

**Comment by kwrodarmer** [comment]
It appears that there are a handful of issues:

1. prefetch is using our SDL from AWS and is getting reference sequences from NCBI instead of AWS, as it should. This is still an open issue under investigation.
2. blastn_vdb opens a connection to another NCBI database which is triggering rate limiting.

The hypothesis is that blastn doesn't need to create this connection, and can be disabled with the environment variable. If so, it should sidestep the problem of the rate limiting, which itself is being addressed independently.

**Comment by taltman** [comment]
Unfortunately I had to stop that instance, as the charges were too high to justify leaving it on.

I relaunched the instance. Since it was allocated a new IP address, so far it seems I haven't raised any hackles over at NCBI; the warning message hasn't come back.

Going back to the reproducible example, I tried running `blastn_vdb` again, with the `vdb-config` setting for accepting AWS charges turned *OFF*. Here is the error message:

``` 
base) ubuntu@ip-172-31-72-183:/media/storage$ time blastn_vdb -db SRR11647690 -query gene.fna > test.tsv
VDB: 2020-06-18T22:20:55 . int: timeout exhausted while creating file within network system module - cannot Make Compute Environment Token
VDB: 2020-06-18T22:21:00 . err: timeout exhausted while creating file within network system module - Cannot KHttpFileTimedReadChunked
VDB: 2020-06-18T22:22:38 . int: timeout exhausted while waiting condition within process system module - Error in VDBManagerOpenDBRead(SRR11647690)
Error: Error opening the following db(s): SRR11647690
```

When I run it with accept charges turned *ON*, there is no such error message. Both of these runs are without setting the environment variable that you mentioned.

```
(base) ubuntu@ip-172-31-72-183:/media/storage$ prefetch SRR11647690

2020-06-18T22:43:21 prefetch.2.10.7 int: timeout exhausted while creating file within network system module - cannot Make Compute Environment Token
2020-06-18T22:43:21 prefetch.2.10.7: 1) Downloading 'SRR11647690'...
2020-06-18T22:43:21 prefetch.2.10.7:  Downloading via HTTPS...
2020-06-18T22:44:13 prefetch.2.10.7:  HTTPS download succeed
2020-06-18T22:44:13 prefetch.2.10.7: 1) 'SRR11647690' was downloaded successfully
2020-06-18T22:44:13 prefetch.2.10.7: 'SRR11647690' has 0 unresolved dependencies

(base) ubuntu@ip-172-31-72-183:/media/storage$ time blastn_vdb -db SRR11647690 -query gene.fna > test.tsv
Critical: (307.10) [NCBI-MESSAGE]  [DISPD#ID2@mweb21] Please make sure your firewall does not block ports 4444-4544 bound for 130.14.29.112
```

When I flip on the env var that you requested (to confirm, it's really with double-underscores?), the firewall message sometimes appears, and sometimes doesn't. It doesn't seem consistent/reproducible.

**Comment by kwrodarmer** [comment]
It looks like you left out the `prefetch` step this time.

I don't have an explanation of the behavior you're seeing. We regularly create compute environment tokens without issue. We'll try to recreate it.

Regarding the env variable, I copied it exactly as it was given to me by the blast support guys. It wasn't guaranteed, but we wanted to see if it helped you out.

**Comment by taltman** [comment]
I actually did run the prefetch steps, but just omitted it from the comment, as it is generally working OK.

There's a whole menagerie of error messages that I'm getting, and I'm not always getting them reproducibly. It might have to do with the amount of requests coming from my instance to various servers at any given moment.

Here are the error messages from above that I don't understand:

- Please make sure your firewall does not block ports 4444-4544 bound for 130.14.29.112
- int: timeout exhausted while creating file within network system module - cannot Make Compute Environment Token
- Error: Cannot get VDB column: /dev/shm/SRR11647688/SRR11647688.sra.SEQUENCE.ACCESSION

Any help with deciphering what I'm doing wrong in those cases would be greatly appreciated. I'll be sure to update this issue as well if I learn how to reproducibly trigger some of them.

**Comment by kwrodarmer** [comment]
You shouldn't have to decipher them...

The first one goes back to blast trying to contact a database inside NCBI. You can't really do more than setting that environment variable as far as I know, but I'll try to get a blast team member to help.

The second one comes from VDB and indicates that it thinks it's operating with AWS but is for some reason timing out when asking AWS to generate a token representing your compute environment. We do this every day without issue.

The last message again comes from blast, so we need to get some help from them.

**Comment by sachalau** [comment]
Hello,

I don't know if its related, but I'm getting errors when trying to prefetch and fasterq-dump on an AWS us-east-1 location too. However the errors are different : 

When doing fasterq-dump directly : 
```
2020-06-19T15:29:30 fasterq-dump.2.10.7 sys: mbedtls_ssl_get_verify_result returned 0x8 (  !! The certificate is not correctly signed by the trusted CA  )
2020-06-19T15:29:30 fasterq-dump.2.10.7 err: connection failed while opening file within cryptographic module - error with https open 'https://locate.ncbi.nlm.nih.gov/sdlr/sdlr.fcgi?jwt=eyJhbGciOiJSUzI1NiIsImtpZCI6InNkbGtpZDEiLCJ0eXAiOiJKV1QifQ.eyJleHAiOjE1OTI1ODQxNzAsImlhdCI6MTU5MjU4MDU3MCwibGluayI6Imh0dHBzOi8vc3JhLXB1Yi1ydW4tMS5zMy5hbWF6b25hd3MuY29tL1NSUjU3MDk4NDgvU1JSNTcwOTg0OC4xP25jYmlfcGhpZD05MzlCODZGOTlGNDg1MzI1MDAwMDQzOTU1QjA2NDM4Ri4xLjEmeC1hbXotcmVxdWVzdC1wYXllcj1yZXF1ZXN0ZXIiLCJyZWdpb24iOiJ1cy1lYXN0LTEiLCJzZXJ2aWNlIjoiczMiLCJzaWduaW5nQWNjb3VudCI6InNyYV9zMyIsInRpbWVvdXQiOjYwMDB9.SEJHTjkgnwGUqOe5uymbDiig8AhSeeUDCEMErBqgYLK4ldvyyrI8B-er9DNiXsPnjp7ZI0yetF1bW5IB8WmIrA2oOxPcX3irL1RCe59C2OfQYzuBLhARgRGD5URNJmp01pykgNalC9VfffhqWge53-wyZPRZf5l2YAS-HKIUvrdPMs6GqZiK8PpdfkWOklL1CRFBg1SKNchYw3_uI1zznrcJ_zN8ScAdTywOJ1BJaBXl5KnGNWy3A_PAsQchzEMjb41l4S_rEjXVc3cZ25DkFrTOxktEPzB-n-9VsJCw0bgan2IcDnHrnZlt3E-uxqnPsb6fpDHzG9Y9RvP1rtHpkg'
```

when trying `prefetch` and then `fasterq-dump`

```

2020-06-19T17:50:52.021+02:00 | 2020-06-19T15:50:52  prefetch.2.10.7 sys: mbedtls_ssl_get_verify_result returned 0x8 (  !!  The certificate is not correctly signed by the trusted CA  )
2020-06-19T15:50:52 prefetch.2.10.7 int: connection failed while opening file within cryptographic module - cannot open remote file: https://locate.ncbi.nlm.nih.gov/sdlr/sdlr.fcgi?jwt=eyJhbGciOiJSUzI1NiIsImtpZCI6InNkbGtpZDEiLCJ0eXAiOiJKV1QifQ.eyJleHAiOjE1OTI1ODU0NTEsImlhdCI6MTU5MjU4MTg1MSwibGluayI6Imh0dHBzOi8vc3JhLXB1Yi1ydW4tMS5zMy5hbWF6b25hd3MuY29tL1NSUjU3MDk4NDgvU1JSNTcwOTg0OC4xP25jYmlfcGhpZD1EMEJEMUJDQTExQTI2Rjk1MDAwMDM3OTVGOUVEMTQ3Ni4xLjEmeC1hbXotcmVxdWVzdC1wYXllcj1yZXF1ZXN0ZXIiLCJyZWdpb24iOiJ1cy1lYXN0LTEiLCJzZXJ2aWNlIjoiczMiLCJzaWduaW5nQWNjb3VudCI6InNyYV9zMyIsInRpbWVvdXQiOjYwMDB9.GrRmYAxtGFmCTd-XVyHXrOpjP2d0Y-a5ZG94pT-vAWMVenk9jybI3jmutKAsTAsl6K8upNmIIraZ2xxxNxrlHxDrpNDQld8FYMeaEA5q9qNtiqpmXzmTlzezquhN12OsYl10OhhlL-2phqQNe7D3eJpqLkov3yXehvph5vG6snw5z5BSzDrd-B6QsiW3aSqQ9L0NuNV3BHT5IlfmSGWId6jB3tqIsTd-_xLHkJBIODowfFGy8NZbz_wj1DVn3fDzTfPi9ZB2Z1WObOEGODhI0ZUn-_pxBQGAuaziyTIhZOgIAld14ihS3SR6Mae7db_oX8S_tkKNCcfB7mUNTcp09w
 2020-06-19T15:50:53 prefetch.2.10.7: 1) Downloading 'SRR5709848'...
2020-06-19T15:50:53 prefetch.2.10.7 sys: mbedtls_ssl_get_verify_result returned 0x8 (  !! The certificate is not correctly signed by the trusted CA  )
2020-06-19T15:50:53 prefetch.2.10.7:  Downloading via HTTPS...
2020-06-19T15:50:53 prefetch.2.10.7 sys: mbedtls_ssl_get_verify_result returned 0x8 (  !! The certificate is not correctly signed by the trusted CA  )
2020-06-19T15:50:53 prefetch.2.10.7 int: connection failed while opening file within cryptographic module - Cannot KClientHttpRequestGET: /scratch/SRR5709848/SRR5709848.sra
2020-06-19T15:50:53 prefetch.2.10.7:  HTTPS download failed
```

I thought this might be related as prefetch is trying to download through HTTPS too which in this case it should'nt.

**Comment by kwrodarmer** [comment]
We notice this when running inside a Docker container and using a non-AWS root certificate file. I don't know what conditions you're running under.

**Comment by kwrodarmer** [comment]
> prefetch is trying to download through HTTPS too which in this case it should'nt.

Why do you say it shouldn't be using https?

**Comment by durbrow** [comment]
> ```
> 2020-06-19T15:50:53 prefetch.2.10.7:  Downloading via HTTPS...
> 2020-06-19T15:50:53 prefetch.2.10.7 sys: mbedtls_ssl_get_verify_result returned 0x8 (  !! The certificate is not correctly signed by the trusted CA  )
> 2020-06-19T15:50:53 prefetch.2.10.7 int: connection failed while opening file within cryptographic module - Cannot KClientHttpRequestGET: /scratch/SRR5709848/SRR5709848.sra
> 2020-06-19T15:50:53 prefetch.2.10.7:  HTTPS download failed
> ```

I have seen this error before when running from inside a docker container. If you are running inside a docker container, this might help:
Mount /etc/pki and /etc/ssl in the container, e.g.
```
% docker run -v ~/output/:/output:rw -v /etc/pki:/etc/pki:ro -v /etc/ssl:/etc/ssl:ro --rm sratoolkit:delite delite_docker.sh SRR000001
```

**Comment by sachalau** [comment]
Sorry for highjacking this thread, mounting the volumes as @durbrow advised fixed my problem !

> Why do you say it shouldn't be using https?

I don't know, some wrong assumption. Sorry for the distraction.

**Comment by kwrodarmer** [comment]
@sachalau - good to hear!

We probably need to create a Wiki page about using Docker images in the cloud.

**Comment by nealfos** [comment]
@durbrow noob here, sorry..

**Comment by durbrow** [comment]
I have never seen that error message before. Can you post the entire command line? From what google showed me (just now), it has something to do with docker parsing the command line. Examples were like paths or file names with spaces that aren't quoted or escaped properly. Without the whole command line, I can't begin to guess.

**Comment by nealfos** [comment]
![image](https://user-images.githubusercontent.com/80130645/110537508-c5903200-80f0-11eb-8b47-71dcfe18473f.png)

**Comment by durbrow** [comment]
Got it. The "..." meant the rest of the command (which I didn't supply because that part would vary depending). The "..." is not itself part of the command.

**Comment by nealfos** [comment]
ok- odd. i can run the command from the ec2 machine:   `% docker run -t --rm -v $PWD:/output:rw -w /output ncbi/sra-tools prefetch SRR10985476`  and it succeeds

but this command fails:  `docker run -v /etc/pki:/etc/pki:ro -v /etc/ssl:/etc/ssl:ro ...` 
with this error: `docker: invalid reference format.
See 'docker run --help'.`

**Comment by durbrow** [comment]
It shouldn't be `docker run -v /etc/pki:/etc/pki:ro -v /etc/ssl:/etc/ssl:ro ...`
It should be `docker run -v /etc/pki:/etc/pki:ro -v /etc/ssl:/etc/ssl:ro` ...
The "..." is not part of the command.
For example: `docker run -t --rm -v /etc/pki:/etc/pki:ro -v /etc/ssl:/etc/ssl:ro -v $PWD:/output:rw -w /output ncbi/sra-tools prefetch SRR10985476`
However, it may be that `/etc/pki` does not exist on your system, not all systems have it (or they may put it elsewhere). Try `ls /etc/pki`

**Comment by durbrow** [comment]
I didn't know galaxy was using our docker file. Good to know.

**Comment by nealfos** [comment]
here's the repo link: https://toolshed.g2.bx.psu.edu/repository?repository_id=cb2d31dfab58ee88

**Comment by taltman** [comment]
@nealfos @durbrow Can you please move your discussion to a separate issue, as it doesn't have a connection with my issue? Thanks!

**Comment by klymenko** [comment]
@taltman, first of all use our latest 2.11.0 release.

`Please make sure your firewall does not block ports 4444-4544 bound for 130.14.29.112`
Please send this question to blast-help@ncbi.nlm.nih.gov

`int: timeout exhausted while creating file within network system module - cannot Make Compute Environment Token`
You mentioned it in 2 different contexts:

1. This one looks like failure to open SRR11647690 - retry your command:
```
blastn_vdb -db SRR11647690 -query gene.fna
VDB: 2020-06-18T22:20:55 . int: timeout exhausted while creating file within network system module - cannot Make Compute Environment Token
VDB: 2020-06-18T22:21:00 . err: timeout exhausted while creating file within network system module - Cannot KHttpFileTimedReadChunked
VDB: 2020-06-18T22:22:38 . int: timeout exhausted while waiting condition within process system module - Error in VDBManagerOpenDBRead(SRR11647690)
```
2. The followed prefetch command prints "HTTPS download succeed", "'SRR11647690' was downloaded successfully", "SRR11647690' has 0 unresolved dependencies" and exited with 0.
It means that prefetch succeed and you can ignore this message.
```
prefetch SRR11647690
2020-06-18T22:43:21 prefetch.2.10.7 int: timeout exhausted while creating file within network system module - cannot Make Compute Environment Token
2020-06-18T22:43:21 prefetch.2.10.7: 1) Downloading 'SRR11647690'...
2020-06-18T22:43:21 prefetch.2.10.7:  Downloading via HTTPS...
2020-06-18T22:44:13 prefetch.2.10.7:  HTTPS download succeed
2020-06-18T22:44:13 prefetch.2.10.7: 1) 'SRR11647690' was downloaded successfully
2020-06-18T22:44:13 prefetch.2.10.7: 'SRR11647690' has 0 unresolved dependencies
```

> blastn_vdb -db SRR11647688/SRR11647688.sra -query gene.fna > test.tsv
> Error: Cannot get VDB column: /dev/shm/SRR11647688/SRR11647688.sra.SEQUENCE.ACCESSION

Here run `vdb-validate SRR11647688/SRR11647688.sra` to make sure the run is not corrupted.
